### ISSUE TITLE
Questions regarding early_stopping_rounds & fitted model best_iteration

### ISSUE THREAD

**Issue by borisRa** [microsoft/LightGBM#4834]
Hi ,

I have three questions : 
1) If I am adding 'train_data_Dataset' to  'valid_sets' like that :
      `lgbm_model = lgbm.train(lgbm_params,
							  train_set = train_data_Dataset,	   					  
							  early_stopping_rounds = 10, 					  
							  valid_sets = [ train_data_Dataset,valid_data_Dataset], #valid_sets = [train_data, valid_data],
							  verbose_eval = 100 ,
							  evals_result = eval_result_dict
							  )`
How the  training data is being ignored  during the early stop functionality  ?

2) I am using the 'n_estimators' that controls number of build decision trees  (for example n_estimators =1000 ). 
   But through the training the num of best_iteration = 20. If I am retraining the  model again **on all data** - I should change the n_estimators  to be 20  (n_estimators  =20)?  

3)  Why in prediction there is 'num_iteration' attribute ? The model uses the best achieved iteration automatically  so why this attributed is needed ?

Thanks,
Boris R

**Comment by jameslamb** [comment]
Thanks very much for using LightGBM!

> How the training data is being ignored during the early stop functionality

Please see https://github.com/microsoft/LightGBM/issues/4679#issuecomment-955872028 for a detailed recent description of how LightGBM ignores the training data for early stopping.

> through the training the num of best_iteration = 20. If I am retraining the model again on all data - I should change the n_estimators to be 20 (n_estimators =20)

Your interpretation is correct. `best_iteration = 20` means "LightGBM stopped training after 21 iterations, because performance on the validation data stopped improving". This suggests that if you re-train on the full dataset with, for example, `n_estimators = 1000`, the training time, computing, and memory consumed for most of those additional training rounds will not make the resulting model much better. Note that `best_iteration` is 0-based, e.g. `best_iteration = 20` means "21 iterations".

If you have a fairly small dataset (say, less than 10,000 observations), I recommend using `n_estimators` slightly higher than `best_iteration` you obtained during training where somme data was held out for validation.

> Why in prediction there is 'num_iteration' attribute?

Tree-based gradient boosting models like LightGBM create predictions by adding the predicted values from each tree, in sequential order. That means that if you have a 10-tree model, creating a prediction on a sample involves passing it through 10 trees and summing the values of the leaf nodes the sample falls into.

If your goal is "generate the most accurate possible prediction, regardless of how expensive it is to compute it", then leaving `num_iteration` at the default value is the right choice.

However, that's not always the case. Consider these scenarios:

- *"I've found that the predictions as of `num_iteration = 500` are not that different than those produced as of `num_iteration = 5000`, so I'm comfortable using fewer iterations in exchange for each prediction being calculated faster"*
- *"I want to inspect the model structure by checking  which leaf nodes some evaluation data fall in, and I just want to visually inspect the output for the first 10 trees"* (using `pred_leaf`, see the `Booster.predict()` docs ([link](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.Booster.html#lightgbm.Booster.predict)))
- *"I want to figure out how different the predictions would be is I only used the first `100` trees vs. the full model I trained, because I'm interested in trying to make the model object smaller or reducing the time-per-prediction"

**Comment by borisRa** [comment]
Thank you very much !
